### PR TITLE
[ASM] - Expander - Service Ownership Model Update

### PR DESCRIPTION
**Pull request by kball-pa**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-hq.paloaltonetworks.local/browse/EXPANDR-4989

## Description
Updates the Service Ownership playbook to use a more sophisticated machine learning model. In particular, the RankServiceOwners script is updated to:

- download the model from GCS and load it
- turn the data in `asmserviceownerunrankedraw` into a numerical array of features
- run inference on the featurized set of potential owners
- normalize this score to a value between 0.5 and 1
- rank them according to this normalized score

## Must have
- [ x] Tests
- [ x] Documentation

### CODE REVIEW

**Comment by kball-pa** [comment]
@anas-yousef @efelmandar are the security issues flagged above (using dill for serialization/de-serialization) blocking issues? for context this automation reads a pickled machine learning model from a remote GCS bucket with privileged write access

**Comment by kball-pa** [comment]
Converted PR to draft to block merging until testing with final prod GCS bucket is done.

Testing is currently blocked on devops work to grant read access to the relevant service accounts.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:459]: If I'm reading it right, this comment is not a comment on the  OwnerFeaturizationPipeline, so much as a process recommendation -- that we should flag for all the relevant teams that by using the "source" property on the `owners` provided upstream, we are coupling this code & model releases tightly to the actual values in that field. And if we can make a better recommendation, we should do so.

So, could you:
1. socialize the flag/concern/recommendation from this paragraph with the relevant teams, and provide a suggestion for a better implementation of the input `owners` format, and figure out who will make the change and by which release, or whether we won't prioritize it at all.
2. make sure there is a note about "if you change the source values, you'll need to update the ML service ownership model". in the document that that team is maintaining around the downstream implications of hardcoded values (I believe @johnnywilkes owns this document re: the other ongoing initiative, and can direct you to it)
3. remove this comment from this PR (because it isn't about using the code at all)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:472]: @kball-pa let's try to resolve this FIXME before merge.

ideally it looks like we'd have RC add a field upstream that is specifically for the "source system" concept. can you look into whether this is feasible on a reasonable timeframe?

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:572]: a lot of the stuff that I originally wrote reads like unreviewed prototype code (because it is), but this REALLY reads like unreviewed prototype code -- is there a more robust way to handle it? maybe instead of a single variable `self.FEATURES` there should be 2 variables? other ideas?

(feel free to fix & review the other code too!)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:26]: this test is dangerous -- if something exceptional happens, we'll end up with the model on disk being `[1,2,3]`, and it will exist and have non-zero size, so all featurizations afterwards will always fail for that tenant, until the disk is somehow cleared.

instead, I'd recommend:
* `load_pickled_xpanse_object` should take the path as an input. the path should NOT be a public constant, and it should not be imported in this testing file
* in the test, you should use Python's `tempfile` library to manage the temporary existence of this file -- it is careful about cleanup under extraordinary conditions, and it clearly marks what files won't persist

and/or don't test this code at all.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:41]: I'm struggling to understand what's being tested here. is the idea that we're passing in arbitrary numbers and they're getting returned in sorted order? if so, given these inputs & outputs, it seems simple -- why is it worth a method in our codebase for this instead of relying on `sorted` (or `list.sort`)? and if sorting is what's going on and it *is* more complex than can be reasonably handled by a passsthrough to `sort`, could you rename the method to be clearer? (maybe capturing that it's a `sort` more than a `rank`, or naming with another characteristic that matters in making a plain `sorted` not appropriate?)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:193]: where did these numbers come from? and the ones in line 215?

If the right value of what we are testing isn't readily apparent to the reader of the test, we should change something about the test -- tests should be verifying obvious functionality. some options might be:
1. break apart the test so the source of the numbers is apparent (maybe into multiple tests -- e.g., maybe we test featurization of the `owners` values separate from model prediction)
2. add enough context so the number source is apparent
3. make assertions about something that *is* apparent from the data (like who is the most / least likely to be the best owner point-of-contact, or that we do get a numeric answer)

you've already got 1, 2, and 3 covered below -- so after reviewing the rest of this doc, I'm wondering, does this test cover something we aren't testing elsewhere? could we simplify it?

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:692]: should this be >= 1, since there is only one obvious match?

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:218]: can this be simplified? the `owners` seem very complex if you just want to verify that if `load_pickled_xpanse_object` raises an exception, we still get an output.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:246]: can this be simplified? the `owners` seem very complex if you just want to verify that if `load_pickled_xpanse_object` raises an exception, we still get an output.

Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:246]: should `test_score_model_inference_fail` be the same as `test_score_model_load_fail`? they look very similar to me -- if they're different, and if even after simplifying they still look very similar, could you call out the difference in what you're trying to test in the docstring?

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (ptoman-pa) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:1]: There's a docstring format for tests that I've had to use in this codebase before -- something about When/Given/Then or the like. you'll probably be asked to refactor to this format (you can find the format in a parallel test file) -- it might be worth confirming this and then getting ahead of it.

**Review by ptoman-pa — COMMENTED** [review]
big PR -- lots of progress!

File: Packs/CortexAttackSurfaceManagement/.pack-ignore
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/.pack-ignore:44]: yeah it's an allow list. not sure about space tokenization; in this case it used dash tokenization -- the release-notes check failed on the docker image tag "xpanse-ml-ev2" (with an added message like "did you mean 'xml'"?) and prompted me to add "ml" to the .pack-ignore. 

I guess i could add the full image tag "xpanse-ml-ev2" to the ignore list?

File: Packs/CortexAttackSurfaceManagement/Scripts/GetProjectOwners/GetProjectOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/GetProjectOwners/GetProjectOwners.py:33]: might be a new demisto standard? with the imported type hints from `typing` the `pre-commit` check in CI/CD failed and outputted a diff removing the imported type hints in favor of the built-in types, and once i applied the diff the validation pass. but that didn't happen on my previous PR's

File: Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md
Comment (BigEasyJ) [reply Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md:7]: It depends. If something would be relevant for a customer or external parties, yes please put why. If it's more relevant internally, we should use internal options or doc strings / comments in the code for "why"?

File: Packs/CortexAttackSurfaceManagement/.pack-ignore
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/.pack-ignore:44]: after discussing with pamela: planning to leave as-is (unless others recommend otherwise)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py]: @BigEasyJ and I discussed the addition of field in `asmserviceowner` for this specific info; ticketed it and John will see if he can get it prioritized for the 2.3 Expander release

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py]: see https://github.com/demisto/content/pull/28375#discussion_r1276594560

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:438]: couldn't come up with anything better but did add an example!

File: Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md:1]: Please undo this change.

File: Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_27.md
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_27.md:5]:
```suggestion
Updated the playbook to pass ASM system IDs as input to the RankServiceOwners automation.
```

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:5]: NIT: Please alphabetize these.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py
Comment (BigEasyJ) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners_test.py:1]: +1 Please you docstrings for all functions.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:46]: Is the a mutable variable? If so I would lowercase it.

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:57]: [Does this work without credentials?](https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.client.Client)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:495]: After meeting we decided this is the best option for now, especially given the change we made to include [`Owner-In-Tags-From-<PLATFORM>` as a source in `asmserviceowner`](https://github.com/demisto/content/blob/master/Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Enrichment.yml#L1208). There is not an easy way to currently pull a unique list of integrations the service owner was found in.

cc: @johnnywilkes @capanw

**Review by BigEasyJ — CHANGES_REQUESTED** [review]
Forgot to hit submit yesterday on my review, but I have added additional things since then as well.

File: Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml
Comment (ptoman-pa) [reply Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:56]: I'm assuming this comment means "it's ok to leave this PR as it is" 👍

File: Packs/CortexAttackSurfaceManagement/.pack-ignore
Comment (ptoman-pa) [reply Packs/CortexAttackSurfaceManagement/.pack-ignore:44]: leaving alone makes sense to me -- adding "ml" to the spellchecker seems very reasonable.

File: Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md
Comment (ptoman-pa) [reply Packs/CortexAttackSurfaceManagement/ReleaseNotes/1_6_26.md]: I think this means "I'm comfortable with these release notes" -- but if you aren't please say so

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (ptoman-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:57]: ^ great catch -- there should be a try/except around the appropriate code, with a meaningful log message if service ownership can't run (it may be all the way out in `main` or otherwise near the beginning of all this functionality)

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:55]: "infrastructure-related service account"

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:321]: i think it's fine!

File: Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py
Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:57]: yes, the invocation is surrounded by a try/except which will assign a fixed fallback score per owner if the model can't be downloaded for any reason

File: Packs/CortexAttackSurfaceManagement/pack_metadata.json
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/pack_metadata.json:1]: Please add the release notes `.md` file.

**Comment by omerKarkKatz** [comment]
Hi @kball-pa, @johnnywilkes, we haven’t heard from you in a while.
What is the status here? 
Do you need any help with the pull request?
Please feel free to reach out to me here or on Slack.
Thanks again for contributing to our repo, hope to hear from you soon.

**Comment by johnnywilkes** [comment]
@omerKarkKatz , we are still working on this.  We will approve within our group and let you know.

**Comment by kball-pa** [comment]
Thanks @johnnywilkes . I've addressed the comments from @BigEasyJ but don't understand the current failures in pre-commit, review_release_notes, and run validations. Lmk if folks see further changes I should make.

File: Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml
Comment (kball-pa) [inline Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:74]: @BigEasyJ don't we actually want to remove lines 74-104?

Comment (kball-pa) [reply Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml]: i went ahead and made that change; please review [this commit](https://github.com/demisto/content/pull/28375/commits/f050426e4f9ac760ceeaae503db74b934def8c72)

**Comment by BigEasyJ** [comment]
There should only be errors on `pull-check` and `release_notes_review` now and those are known issues that you can ignore.

File: Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml]: Does your script still not need to check for `alert.asmsystemids.type` have the value `ASSET-NAME` or `ASSET-ID`?

File: Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml
Comment (BigEasyJ) [inline Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml:55]:
```suggestion
        complex:
          root: alert.asmserviceownerunrankedraw
```

**Review by BigEasyJ — CHANGES_REQUESTED** [review]

File: Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml
Comment (BigEasyJ) [reply Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_Service_Ownership.yml]: Discussed: `ASSET-ID` is not needed. `ASSET-NAME` is used for the scripts business logic optionally.

**Review by BigEasyJ — APPROVED** [review]
Looks good, thanks for all of the work and collaboration!

**Review by omerKarkKatz — APPROVED** [review]

**Review by omerKarkKatz — APPROVED** [review]

**Pull request merged**
